### ISSUE TITLE
Getting OpenCL to work on linux Qualcomm 820

### ISSUE THREAD

**Issue by pgaston** [XiaoMi/mace#579]
Before you open an issue, please make sure you have tried the following steps:

1. Make sure your **environment** is the same with (https://mace.readthedocs.io/en/latest/installation/env_requirement.html).   Using mace-dev
2. Have you ever read the document for your usage?   Yes
3. Check if your issue appears in [HOW-TO-DEBUG](https://mace.readthedocs.io/en/latest/development/how_to_debug.html) or [FAQ](https://mace.readthedocs.io/en/latest/faq.html).
4. The form below must be filled.

------------------------

### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Platform - Ubuntu 18.04
Docker - mace-dev (full), new image
Target - Qualcomm 820
      Kernel: Linux 3.18.71-perf
      Architecture: arm64
- **NDK version(e.g., 15c)**:  n.a.
- **GCC version(if compiling for host, e.g., 5.4.0)**:
docker - gcc (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
target - gcc (GCC) 4.9.3
- **MACE version (Use the command: git describe --long --tags)**:
v0.12.0-0-ga610d50
- **Python version(2.7)**: 
docker - Python 2.7.12
target - Python 2.7.9
- **Bazel version (e.g., 0.13.0)**:
n.a., but 0.16.0
### Model deploy file (*.yml)
```
library_name: mobilenet-v1
target_abis: ['aarch64']
target_socs: [MSM8996]
model_graph_format: file
model_data_format: file
models:
  mobilenet_v1:
    platform: tensorflow
    model_file_path: ../../mace-models/mobilenet-v1/mobilenet-v1-1.0.pb
    #model_file_path: https://cnbj1.fds.api.xiaomi.com/mace/miai-models/mobilenet-v1/mobilenet-v1-1.0.pb
    model_sha256_checksum: 71b10f540ece33c49a7b51f5d4095fc9bd78ce46ebf0300487b2ee23d71294e6
    subgraphs:
      - input_tensors:
          - input
        input_shapes:
          - 1,224,224,3
        output_tensors:
          - MobilenetV1/Predictions/Reshape_1
        output_shapes:
          - 1,1001
    runtime: gpu
```

### Describe the problem
Runs fine targeting cpu.    (51ms on target)

Fails targeting gpu.   Shows "The device does not support OpenCL" when I try to run.

However, I've used OpenCL on this target using opencv dnn - which works (although incredibly slow.)    I've also used ARM CL which used the GPU.

### To Reproduce
Steps to reproduce the problem:
```
Following instructions in tutorial to the letter (I think...)
1. cd container/path/mace
2. tools/clear_workspace.sh
3. RUNTIME=GPU bash tools/cmake/cmake-build-aarch64-linux-gnu.sh
4. python tools/python/convert.py \
  --config ../../mace-models/mobilenet-v1/mobilenet-v1.yml
5. python tools/python/run_model.py  \
    --config ../../mace-models/mobilenet-v1/mobilenet-v1.yml \
    --validate \
    --target_abi aarch64-linux-gnu \
    --target_socs root@192.168.8.1 \
    --device_conf ../../mace-models/820.yml
```

### Error information / logs
Please include the **full** log and/or traceback here.
Repeats the 9 lines at the end indefinitely...
```
/container/path/mace/mace# python tools/python/run_model.py      --config ../../mace-models/mobilenet-v1/mobilenet-v1.yml     --benchmark     --target_abi aarch64-linux-gnu     --target_socs root@192.168.8.1     --device_conf ../../mace-models/820.yml
CMD> bazel build //mace/proto:mace_py
WARNING: detected http_proxy set in env, setting no_proxy for localhost.
WARNING: --batch mode is deprecated. Please instead explicitly shut down your Bazel server using the command "bazel shutdown".
Loading:
Loading: 0 packages loaded
Analyzing: target //mace/proto:mace_py (12 packages loaded)
INFO: Analysed target //mace/proto:mace_py (17 packages loaded).
INFO: Found 1 target...
[0 / 19] [-----] BazelWorkspaceStatusAction stable-status.txt
Target //mace/proto:mace_py up-to-date:
bazel-genfiles/mace/proto/mace_pb2.py
INFO: Elapsed time: 2.078s, Critical Path: 0.10s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action

CMD> cp -f bazel-genfiles/mace/proto/mace_pb2.py /container/path/mace/mace/tools/python/py_proto

CMD> bazel build //third_party/caffe:caffe_py
WARNING: detected http_proxy set in env, setting no_proxy for localhost.
WARNING: --batch mode is deprecated. Please instead explicitly shut down your Bazel server using the command "bazel shutdown".
Loading:
Loading: 0 packages loaded
Analyzing: target //third_party/caffe:caffe_py (12 packages loaded)
INFO: Analysed target //third_party/caffe:caffe_py (17 packages loaded).
INFO: Found 1 target...
[2 / 12] [-----] BazelWorkspaceStatusAction stable-status.txt
Target //third_party/caffe:caffe_py up-to-date:
bazel-genfiles/third_party/caffe/caffe_pb2.py
INFO: Elapsed time: 1.933s, Critical Path: 0.06s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action

CMD> cp -f bazel-genfiles/third_party/caffe/caffe_pb2.py /container/path/mace/mace/tools/python/py_proto

!!! {'Snap820': {'target_abis': ['aarch64'], 'models': 'Qualcomm 820', 'username': 'root', 'target_socs': ['MSM8996'], 'address': '192.168.8.1'}}
tools/python/run_model.py:73: Run on devices: ['root@192.168.8.1']
tools/python/run_model.py:84: Run model mobilenet_v1
{'model_sha256_checksum': '71b10f540ece33c49a7b51f5d4095fc9bd78ce46ebf0300487b2ee23d71294e6', 'output_shapes': ['1,1001'], 'output_tensors': ['MobilenetV1/Predictions/Reshape_1'], 'input_tensors': ['input'], 'input_shapes': ['1,224,224,3'], 'platform': 'tensorflow', 'model_file_path': '../../mace-models/mobilenet-v1/mobilenet-v1-1.0.pb', 'runtime': 'gpu'}
{'output_data_types': [1], 'model_sha256_checksum': '71b10f540ece33c49a7b51f5d4095fc9bd78ce46ebf0300487b2ee23d71294e6', 'data_type': 3, 'input_ranges': [[-1.0, 1.0]], 'input_data_formats': [<DataFormat.NHWC: 1>], 'output_shapes': [[1, 1001]], 'output_tensors': ['MobilenetV1/Predictions/Reshape_1'], 'input_tensors': ['input'], 'input_shapes': [[1, 224, 224, 3]], 'platform': <Platform.TENSORFLOW: 0>, 'output_data_formats': [<DataFormat.NHWC: 1>], 'model_file_path': '../../mace-models/mobilenet-v1/mobilenet-v1-1.0.pb', 'input_data_types': [1], 'runtime': <DeviceType.GPU: 2>}
CMD> ssh root@192.168.8.1 mkdir -p /tmp/mace_run/mobilenet_v1/interior

CMD> ssh root@192.168.8.1 mkdir -p /tmp/mace_run/mobilenet_v1

CMD> scp -r build/mobilenet_v1/model/mobilenet_v1.pb root@192.168.8.1:/tmp/mace_run/mobilenet_v1

CMD> ssh root@192.168.8.1 mkdir -p /tmp/mace_run/mobilenet_v1

CMD> scp -r build/mobilenet_v1/model/mobilenet_v1.data root@192.168.8.1:/tmp/mace_run/mobilenet_v1

tools/python/run_model.py:264: Generate input file: /tmp/tmpfyjhv5/mobilenet_v1_input
CMD> ssh root@192.168.8.1 mkdir -p /tmp/mace_run/mobilenet_v1/validate_in

CMD> scp -r /tmp/tmpfyjhv5 root@192.168.8.1:/tmp/mace_run/mobilenet_v1/validate_in

CMD> ssh root@192.168.8.1 mkdir -p /tmp/mace_run/mobilenet_v1/validate_out

!!! {'Snap820': {'target_abis': ['aarch64'], 'models': 'Qualcomm 820', 'username': 'root', 'target_socs': ['MSM8996'], 'address': '192.168.8.1'}}
Run on devices: ['root@192.168.8.1']
Install target from build/cmake-build/aarch64-linux-gnu/install/bin/mace_run to /tmp/mace_run/mobilenet_v1
CMD> ssh root@192.168.8.1 mkdir -p /tmp/mace_run/mobilenet_v1

CMD> scp -r build/cmake-build/aarch64-linux-gnu/install/bin/mace_run root@192.168.8.1:/tmp/mace_run/mobilenet_v1

['', 'MACE_INTERNAL_STORAGE_PATH=/tmp/mace_run/mobilenet_v1/interior', 'LD_LIBRARY_PATH=/tmp/mace_run/mobilenet_v1', '/tmp/mace_run/mobilenet_v1/mace_run', '--device=GPU', '--input_data_format=NHWC', '--output_node=MobilenetV1/Predictions/Reshape_1', '--output_shape=1,1001', '--model_data_file=/tmp/mace_run/mobilenet_v1/mobilenet_v1.data', '--input_node=input', '--model_file=/tmp/mace_run/mobilenet_v1/mobilenet_v1.pb', '--model_name=mobilenet_v1', '--output_data_format=NHWC', '--input_shape=1,224,224,3', '--benchmark', '--input_file=/tmp/mace_run/mobilenet_v1/validate_in/mobilenet_v1', '--output_file=/tmp/mace_run/mobilenet_v1/validate_out/mobilenet_v1']
 MACE_INTERNAL_STORAGE_PATH=/tmp/mace_run/mobilenet_v1/interior LD_LIBRARY_PATH=/tmp/mace_run/mobilenet_v1 /tmp/mace_run/mobilenet_v1/mace_run --device=GPU --input_data_format=NHWC --output_node=MobilenetV1/Predictions/Reshape_1 --output_shape=1,1001 --model_data_file=/tmp/mace_run/mobilenet_v1/mobilenet_v1.data --input_node=input --model_file=/tmp/mace_run/mobilenet_v1/mobilenet_v1.pb --model_name=mobilenet_v1 --output_data_format=NHWC --input_shape=1,224,224,3 --benchmark --input_file=/tmp/mace_run/mobilenet_v1/validate_in/mobilenet_v1 --output_file=/tmp/mace_run/mobilenet_v1/validate_out/mobilenet_v1
Runing ...
['', 'MACE_INTERNAL_STORAGE_PATH=/tmp/mace_run/mobilenet_v1/interior', 'LD_LIBRARY_PATH=/tmp/mace_run/mobilenet_v1', '/tmp/mace_run/mobilenet_v1/mace_run', '--device=GPU', '--input_data_format=NHWC', '--output_node=MobilenetV1/Predictions/Reshape_1', '--output_shape=1,1001', '--model_data_file=/tmp/mace_run/mobilenet_v1/mobilenet_v1.data', '--input_node=input', '--model_file=/tmp/mace_run/mobilenet_v1/mobilenet_v1.pb', '--model_name=mobilenet_v1', '--output_data_format=NHWC', '--input_shape=1,224,224,3', '--benchmark', '--input_file=/tmp/mace_run/mobilenet_v1/validate_in/mobilenet_v1', '--output_file=/tmp/mace_run/mobilenet_v1/validate_out/mobilenet_v1']
CMD> ssh root@192.168.8.1  MACE_INTERNAL_STORAGE_PATH=/tmp/mace_run/mobilenet_v1/interior LD_LIBRARY_PATH=/tmp/mace_run/mobilenet_v1 /tmp/mace_run/mobilenet_v1/mace_run --device=GPU --input_data_format=NHWC --output_node=MobilenetV1/Predictions/Reshape_1 --output_shape=1,1001 --model_data_file=/tmp/mace_run/mobilenet_v1/mobilenet_v1.data --input_node=input --model_file=/tmp/mace_run/mobilenet_v1/mobilenet_v1.pb --model_name=mobilenet_v1 --output_data_format=NHWC --input_shape=1,224,224,3 --benchmark --input_file=/tmp/mace_run/mobilenet_v1/validate_in/mobilenet_v1 --output_file=/tmp/mace_run/mobilenet_v1/validate_out/mobilenet_v1
I /container/path/mace/mace/mace/tools/mace_run.cc:527] model name: mobilenet_v1
I /container/path/mace/mace/mace/tools/mace_run.cc:528] mace version: v0.12.0-0-ga610d50
I /container/path/mace/mace/mace/tools/mace_run.cc:529] input node: input
I /container/path/mace/mace/mace/tools/mace_run.cc:530] input shape: 1,224,224,3
I /container/path/mace/mace/mace/tools/mace_run.cc:531] output node: MobilenetV1/Predictions/Reshape_1
I /container/path/mace/mace/mace/tools/mace_run.cc:532] output shape: 1,1001
I /container/path/mace/mace/mace/tools/mace_run.cc:533] input_file: /tmp/mace_run/mobilenet_v1/validate_in/mobilenet_v1
I /container/path/mace/mace/mace/tools/mace_run.cc:534] output_file: /tmp/mace_run/mobilenet_v1/validate_out/mobilenet_v1
I /container/path/mace/mace/mace/tools/mace_run.cc:535] input dir:
I /container/path/mace/mace/mace/tools/mace_run.cc:536] output dir: output
I /container/path/mace/mace/mace/tools/mace_run.cc:537] model_data_file: /tmp/mace_run/mobilenet_v1/mobilenet_v1.data
I /container/path/mace/mace/mace/tools/mace_run.cc:538] model_file: /tmp/mace_run/mobilenet_v1/mobilenet_v1.pb
I /container/path/mace/mace/mace/tools/mace_run.cc:539] device: GPU
I /container/path/mace/mace/mace/tools/mace_run.cc:540] round: 1
I /container/path/mace/mace/mace/tools/mace_run.cc:541] restart_round: 1
I /container/path/mace/mace/mace/tools/mace_run.cc:542] gpu_perf_hint: 3
I /container/path/mace/mace/mace/tools/mace_run.cc:543] gpu_priority_hint: 3
I /container/path/mace/mace/mace/tools/mace_run.cc:544] omp_num_threads: -1
I /container/path/mace/mace/mace/tools/mace_run.cc:545] cpu_affinity_policy: 1
I /container/path/mace/mace/mace/libmace/mace.cc:464] Creating MaceEngine, MACE version: v0.12.0-0-ga610d50
I /container/path/mace/mace/mace/libmace/mace.cc:503] Initializing MaceEngine
I /container/path/mace/mace/mace/libmace/mace.cc:636] Destroying MaceEngine
I /container/path/mace/mace/mace/tools/mace_run.cc:596] restart round 0
I /container/path/mace/mace/mace/libmace/mace.cc:911] Create MaceEngine from model graph proto and weights data
I /container/path/mace/mace/mace/libmace/mace.cc:464] Creating MaceEngine, MACE version: v0.12.0-0-ga610d50
E /container/path/mace/mace/mace/core/runtime/opencl/opencl_wrapper.cc:327] Failed to load OpenCL library, please make sure there exists OpenCL library on your device, and your APP have right to access the library.
E /container/path/mace/mace/mace/core/runtime/opencl/opencl_runtime.cc:278] No OpenCL platforms found
I /container/path/mace/mace/mace/libmace/mace.cc:503] Initializing MaceEngine
W /container/path/mace/mace/mace/libmace/mace.cc:58] The device does not support OpenCL
I /container/path/mace/mace/mace/libmace/mace.cc:507] CheckGPUAvalibility(net_def, device_.get()) failed with error: Out of resources
E /container/path/mace/mace/mace/tools/mace_run.cc:265] Create engine runtime error, retry ... errcode: Out of resources: Out of resources: Out of resources: Out of resources
I /container/path/mace/mace/mace/libmace/mace.cc:911] Create MaceEngine from model graph proto and weights data
I /container/path/mace/mace/mace/libmace/mace.cc:464] Creating MaceEngine, MACE version: v0.12.0-0-ga610d50
E /container/path/mace/mace/mace/core/runtime/opencl/opencl_runtime.cc:278] No OpenCL platforms found
I /container/path/mace/mace/mace/libmace/mace.cc:636] Destroying MaceEngine
I /container/path/mace/mace/mace/libmace/mace.cc:503] Initializing MaceEngine
W /container/path/mace/mace/mace/libmace/mace.cc:58] The device does not support OpenCL
I /container/path/mace/mace/mace/libmace/mace.cc:507] CheckGPUAvalibility(net_def, device_.get()) failed with error: Out of resources
E /container/path/mace/mace/mace/tools/mace_run.cc:265] Create engine runtime error, retry ... errcode: Out of resources: Out of resources: Out of resources: Out of resources
I /container/path/mace/mace/mace/libmace/mace.cc:911] Create MaceEngine from model graph proto and weights data
I /container/path/mace/mace/mace/libmace/mace.cc:464] Creating MaceEngine, MACE version: v0.12.0-0-ga610d50
E /container/path/mace/mace/mace/core/runtime/opencl/opencl_runtime.cc:278] No OpenCL platforms found
I /container/path/mace/mace/mace/libmace/mace.cc:636] Destroying MaceEngine
I /container/path/mace/mace/mace/libmace/mace.cc:503] Initializing MaceEngine
W /container/path/mace/mace/mace/libmace/mace.cc:58] The device does not support OpenCL
I /container/path/mace/mace/mace/libmace/mace.cc:507] CheckGPUAvalibility(net_def, device_.get()) failed with error: Out of resources
E /container/path/mace/mace/mace/tools/mace_run.cc:265] Create engine runtime error, retry ... errcode: Out of resources: Out of resources: Out of resources: Out of resources
I /container/path/mace/mace/mace/libmace/mace.cc:911] Create MaceEngine from model graph proto and weights data
I /container/path/mace/mace/mace/libmace/mace.cc:464] Creating MaceEngine, MACE version: v0.12.0-0-ga610d50
E /container/path/mace/mace/mace/core/runtime/opencl/opencl_runtime.cc:278] No OpenCL platforms found

```

### Additional context
Device.yml (820.yml)
```
devices:
  Snap820:
    target_abis: [aarch64]
    # device soc - https://en.wikipedia.org/wiki/List_of_Qualcomm_Snapdragon_systems-on-chip#Snapdragon_820_and_821_(2016)
    target_socs: [MSM8996]
    # not sure of this
    models: Qualcomm 820
    address: 192.168.8.1
    username: root
```
and
```
# opkg print-architecture
arch all 1
arch any 6
arch noarch 11
arch aarch64 16
arch apq8096 21
arch arm 26
arch armv4 31
arch armv5 36
arch armv5-vfp 41
arch armv5e 46
arch armv5e-vfp 51
arch armv6 56
arch armv6-vfp 61
arch armv7a 66
arch armv7a-vfp 71
arch armv7a-vfp-neon 76
```
and plenty of disk space available:
```
# df
Filesystem     1K-blocks      Used Available Use% Mounted on
overlay        982940092 406984196 525955564  44% /
tmpfs              65536         0     65536   0% /dev
tmpfs            8016176         0   8016176   0% /sys/fs/cgroup
shm                65536         0     65536   0% /dev/shm
/dev/nvme0n1p2 982940092 406984196 525955564  44% /container/path
udev             7992060         0   7992060   0% /dev/bus/usb
tmpfs            1603236      2244   1600992   1% /run/docker.sock

```

**Comment by pgaston** [comment]
And, using mace-run --device=DSP   worked as well.    (execution time went to 45ms!)

So CPU and DSP work.    GPU does not.   

Thanks!

**Comment by lu229** [comment]
@pgaston , for aarch64, mace will find the opencl so in the following paths:
    // Qualcomm Adreno with Android
    "/system/vendor/lib64/libOpenCL.so",
    "/system/lib64/libOpenCL.so",
    // Mali with Android
    "/system/vendor/lib64/egl/libGLES_mali.so",
    "/system/lib64/egl/libGLES_mali.so",
    // Typical Linux board
    "/usr/lib/aarch64-linux-gnu/libOpenCL.so",
it seems that there is no libOpenCL.so in /system/vendor/lib64/ and /system/lib64/, or mace have no permission to load it, you can find where is the libOpenCL.so on you device, if it is a path in common use, we will add it to the mace.

**Comment by pgaston** [comment]
So there is a libOpenCL.so on the system - it's in /usr/lib .   (See below, but this is a 32bit compiled library I believe.)    I tried copying the existing /usr/lib/libOpenCL.so to /usr/lib64 and (the existing) /usr/aarch64-oe-linux and created a new directory  /usr/aarch /usr/aarch64-linux-gnu/ , /system/lib64, etc.  but results were the same.   Not sure if mace even tried to open that file and failed, or whatever.

I then downloaded a 64 bit armv8 libOpenCL.so from deb - https://packages.debian.org/buster/arm64/ocl-icd-libopencl1/download .     This approach had worked for me before getting a 64 bit libstdc++.so.6      I then copied it to all the standard locations.    However, in this case no change - No OpenCL platforms found.    I even put the string "# LD_LIBRARY_PATH=/usr/lib64 ./mace_run ....." to try and force going to that new libOpenCL.so - but no changes.

Any thoughts?    Like is there a simple way to check if OpenCL is available (64 bits?)

This gets back to my previous problem (linking in libstdc++.so.6) in that this platform was compiled for a 32 bit architecture using gcc 4.9 while your compiler is 64 bit gcc 5.4.    

This is error - this is repeated infinite times...
```
I /container/path/mace/mace/mace/libmace/mace.cc:911] Create MaceEngine from model graph proto and weights data
I /container/path/mace/mace/mace/libmace/mace.cc:464] Creating MaceEngine, MACE version: v0.12.0-0-ga610d50
E /container/path/mace/mace/mace/core/runtime/opencl/opencl_runtime.cc:278] No OpenCL platforms found
I /container/path/mace/mace/mace/libmace/mace.cc:636] Destroying MaceEngine
I /container/path/mace/mace/mace/libmace/mace.cc:503] Initializing MaceEngine
W /container/path/mace/mace/mace/libmace/mace.cc:58] The device does not support OpenCL
I /container/path/mace/mace/mace/libmace/mace.cc:507] CheckGPUAvalibility(net_def, device_.get()) failed with error: Out of resources
E /container/path/mace/mace/mace/tools/mace_run.cc:265] Create engine runtime error, retry ... errcode: Out of resources: Out of resources: Out of resources: Out of resources
```

**Comment by lu229** [comment]
@pgaston It seems that you can't use aarch64 but the armhf. For the ABI of armhf, the path for Mac to find libopencl.so is as follows:
// Qualcomm Adreno with Android
"/system/vendor/lib/libOpenCL.so",
"/system/lib/libOpenCL.so",
// Mali with Android
"/system/vendor/lib/egl/libGLES_mali.so",
"/system/lib/egl/libGLES_mali.so",
// Typical Linux board
"/usr/lib/arm-linux-gnueabihf/libOpenCL.so",
You can modify the source code to add /usr/lib to the above array, which is in the file /mace/core/runtime/OpenCL/opencl_wrapper.cc.

**Comment by pgaston** [comment]
**Success** - at least somewhat!!!    MACE - both the tool and your support is excellent btw!

Mobilenet v1 - 13.8 ms
Yolov3 - 750 ms
my model - 861 - but Upsampling isn't supported so there was cpu fallback - I've fixed this before for onnx

Anyway, the funkiness here is on my side - my 820 board is running a 32bit driver for OpenCL.   Further, from the vendor they provided specific CXX Flags.    

Here's what I changed:
- Using the mace-dev docker...
- downloaded older compiler and changed the toolchain to:  (note soft floating point)
set(CMAKE_C_COMPILER "/usr/bin/arm-linux-gnueabi-gcc-4.9")
set(CMAKE_CXX_COMPILER "/usr/bin/arm-linux-gnueabi-g++-4.9")
set(CMAKE_CXX_FLAGS "-march=armv7-a -mfloat-abi=softfp -mfpu=neon-vfpv4 ${CMAKE_CXX_FLAGS}")
- make device yml correct
    target_abis: ['armv7a']
    target_socs: [APQ8096]
- libOpenCL was found without changing anything
